File: storage/db-core/src/main/java/com/coffee/gu/cancel/CancelEntity.java
package com.coffee.gu.cancel;


import com.coffee.gu.Principal;
import com.coffee.gu.enums.PrincipalType;
import com.coffee.gu.BaseEntity;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Table;

import java.math.BigDecimal;
import java.time.OffsetDateTime;

@Entity
@Table(name = "cancel")
public class CancelEntity extends BaseEntity {
    private String principalKey;
    @Enumerated(EnumType.STRING)
    private PrincipalType principalType;
    private String orderKey;
    private Long paymentId;
    private BigDecimal originalAmount;
    private Long issuedCouponId;
    private BigDecimal couponDiscount;
    private BigDecimal paidAmount;
    private BigDecimal canceledAmount;
    private String externalCancelKey;
    private OffsetDateTime canceledAt;

    public CancelEntity() {}

    public CancelEntity(String principalKey, PrincipalType principalType, String orderKey, Long paymentId, BigDecimal originalAmount, Long issuedCouponId, BigDecimal couponDiscount, BigDecimal paidAmount, BigDecimal canceledAmount, String externalCancelKey, OffsetDateTime canceledAt) {
        this.principalKey = principalKey;
        this.principalType = principalType;
        this.orderKey = orderKey;
        this.paymentId = paymentId;
        this.originalAmount = originalAmount;
        this.issuedCouponId = issuedCouponId;
        this.couponDiscount = couponDiscount;
        this.paidAmount = paidAmount;
        this.canceledAmount = canceledAmount;
        this.externalCancelKey = externalCancelKey;
        this.canceledAt = canceledAt;
    }

    public static CancelEntity from(Cancel cancel) {
        return new CancelEntity(
                cancel.getPrincipal().getKey(),
                cancel.getPrincipal().getType(),
                cancel.getOrderKey(),
                cancel.getPaymentId(),
                cancel.getOriginalAmount(),
                cancel.getIssuedCouponId(),
                cancel.getCouponDiscount(),
                cancel.getPaidAmount(),
                cancel.getCanceledAmount(),
                cancel.getExternalCancelKey(),
                cancel.getCanceledAt()
        );
    }

    public Cancel toModel() {
        return new Cancel(
                this.id,
                new Principal(this.principalKey, this.principalType),
                this.orderKey,
                this.paymentId,
                this.originalAmount,
                this.issuedCouponId,
                this.couponDiscount,
                this.paidAmount,
                this.canceledAmount,
                this.externalCancelKey,
                this.canceledAt
        );
    }

    public OffsetDateTime getCanceledAt() {
        return canceledAt;
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/cancel/CancelEntity.kt
package com.coffee.gu.cancel

import com.coffee.gu.BaseEntity
import com.coffee.gu.Principal
import com.coffee.gu.enums.PrincipalType
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.math.BigDecimal
import java.time.OffsetDateTime

@Entity
@Table(name = "cancel")
class CancelEntity @JvmOverloads constructor(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0,
    val principalKey: String,
    @Enumerated(EnumType.STRING)
    val principalType: PrincipalType,
    val orderKey: String,
    val paymentId: Long,
    val originalAmount: BigDecimal,
    val issuedCouponId: Long? = null,
    val couponDiscount: BigDecimal? = null,
    val paidAmount: BigDecimal,
    val canceledAmount: BigDecimal,
    val externalCancelKey: String? = null,
    val canceledAt: OffsetDateTime? = null,
) : BaseEntity() {

    fun toModel(): Cancel {
        return Cancel(
            id = id,
            principal = Principal(principalKey, principalType),
            orderKey = orderKey,
            paymentId = paymentId,
            originalAmount = originalAmount,
            issuedCouponId = issuedCouponId,
            couponDiscount = couponDiscount,
            paidAmount = paidAmount,
            canceledAmount = canceledAmount,
            externalCancelKey = externalCancelKey,
            canceledAt = canceledAt,
        )
    }

    companion object {
        @JvmStatic
        fun from(cancel: Cancel): CancelEntity = CancelEntity(
            id = cancel.id,
            principalKey = cancel.principal.key,
            principalType = cancel.principal.type,
            orderKey = cancel.orderKey,
            paymentId = cancel.paymentId,
            originalAmount = cancel.originalAmount,
            issuedCouponId = cancel.issuedCouponId,
            couponDiscount = cancel.couponDiscount,
            paidAmount = cancel.paidAmount,
            canceledAmount = cancel.canceledAmount,
            externalCancelKey = cancel.externalCancelKey,
            canceledAt = cancel.canceledAt,
        )
    }
}
